test/apps/service/apps/resources_test.clj

```diff
@@ -0,0 +1,78 @@
+(ns apps.service.apps.resources-test
+  (:require [apps.service.apps.de.jobs.resources :as r]
+            [clojure.test :refer [deftest] :as t]))
+
+(def gib (* 1024 1024 1024))
+
+(defn reqs
+  [min-mem max-mem min-cpu max-cpu]
+  {:min_memory_limit (when min-mem (* min-mem gib))
+   :memory_limit     (when max-mem (* max-mem gib))
+   :min_cpu_cores    min-cpu
+   :max_cpu_cores    max-cpu})
+
+(def test-cases
+  [{:container    (reqs 2 256 2 256)
+    :requirements (reqs 4 32 4 16)
+    :expected     (reqs 4 32 4 16)
+    :desc         "all values specified and within range"}
+   {:container    (reqs 2 256 2 256)
+    :requirements (reqs nil 32 4 16)
+    :expected     (reqs 2 32 4 16)
+    :desc         "unspecified minimum memory setting"}
+   {:container    (reqs 2 256 2 256)
+    :requirements (reqs 4 nil 4 16)
+    :expected     (reqs 4 4 4 16)
+    :desc         "unspecified maximum memory setting"}
+   {:container    (reqs 2 256 2 256)
+    :requirements (reqs 4 32 nil 16)
+    :expected     (reqs 4 32 2 16)
+    :desc         "unspecified min cpu setting"}
+   {:container    (reqs 2 256 2 256)
+    :requirements (reqs 4 32 4 nil)
+    :expected     (reqs 4 32 4 4)
+    :desc         "unspecified max cpu setting"}
+   {:container    (reqs nil 256 2 256)
+    :requirements (reqs 4 32 4 16)
+    :expected     (reqs 4 32 4 16)
+    :desc         "unspecified min memory setting in container"}
+   {:container    (reqs 2 nil 2 256)
+    :requirements (reqs 4 32 4 16)
+    :expected     (reqs 4 16 4 16)
+    :desc         "unspecified max memory setting in container, request higher than overall default"}
+   {:container    (reqs 2 256 nil 256)
+    :requirements (reqs 4 32 4 16)
+    :expected     (reqs 4 32 4 16)
+    :desc         "unspecified min cpu setting in container"}
+   {:container    (reqs 2 256 2 nil)
+    :requirements (reqs 4 32 4 16)
+    :expected     (reqs 4 32 4 4)
+    :desc         "unspecified max cpu setting in container, request higher than overall default"}
+   {:container    (reqs 2 256 2 256)
+    :requirements (reqs 1 32 1 16)
+    :expected     (reqs 2 32 2 16)
+    :desc         "minimum requests less than container minimums"}
+   {:container    (reqs 2 256 2 256)
+    :requirements (reqs 4 512 4 512)
+    :expected     (reqs 4 256 4 256)
+    :desc         "maximum requests greater than container maximums"}
+   {:container    (reqs 2 256 2 256)
+    :requirements (reqs 4 nil 4 nil)
+    :expected     (reqs 4 4 4 4)
+    :desc         "maximum requests not specified"}
+   {:container    (reqs 2 256 2 256)
+    :requirements (reqs nil nil nil nil)
+    :expected     (reqs 2 2 2 2)
+    :desc         "no requests specified"}
```

Comment:
Should we add a test case for the issue that #275 intended to fix?

If the container has no minimum requirement, and the client does not submit anything for that min or max requirement, then this service was supposed to submit `0` (or `nil`?) to the jobs services, to allow them to use their own defaults for that requirement.

The default can be different for VICE and condor jobs. For example, VICE will use `8GiB` for its default max memory when nothing is requested, but condor jobs will not submit any value for max memory (allowing the job to use all available memory on the node?). They might be using the same default of `4` for max CPUs.

```clojure
   {:container    (reqs nil 256 nil 256)
    :requirements (reqs nil nil nil nil)
    :expected     (reqs 0 0 0 0)
    :desc         "no requests specified, and no container minimums specified"}
```

Or do we now prefer the apps service to enforce a default max if nothing is requested from the client?

Reply:
Ah, that's a good question. I should probably support this test case as well. I'll take a look at this a little later.